iFixit/react-commerce
# Sync browser url query with collection search parameters

This PR closes #84 

Syncing search params with browser url query required to simplify the filter state logic. Moreover to improve management of side effects the search logic has been refactored using a state machine:

<img width="1059" alt="Screen Shot 2021-09-30 at 18 38 40" src="https://user-images.githubusercontent.com/4640135/135495954-0fca6add-7dc7-4208-9e36-09f12fd1f7d1.png">

> Tip: When using VSCode the machine can be visualized with [this extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode)

This change helped also to reduce re-renders and avoid the initial search when the collection is server side rendered. 
Since `@xstate/fsm` supports only a subset of the full XState library, some of the missing features that are required for this PR have been implemented under `lib/fsm-utils`. This way we get the benefit of modeling state with xstate without the bundle size cost of the full library.

> Note: the search error state has emerged as an edge case that we're currently not handling, but I decided to handle it in a separate PR

## QA

1. Visit the Vercel PR preview
2. Visit the parts collection page
3. Verify that when filters are used the url is updated
4. Verify that when searching for something the url is updated
5. Verify that when the pagination is used the url is updated
6. Verify that search parameters persist when the page is reloaded
